### ISSUE TITLE
How to opt out of !Tag syntax for enums in 0.9.0+?

### ISSUE THREAD

**Issue by willfindlay** [dtolnay/serde-yaml#298]
Hi, I was wondering if there is a way (besides just not upgrading the crate version) to opt out of the `!Tag` syntax for enums and keep the old behaviour?

**Comment by teozkr** [comment]
I'm wondering the same. Tags don't roundtrip cleanly to JSON, aren't compatible with the old format, and don't compose.

**Comment by dtolnay** [comment]
You would need to use https://docs.rs/serde_yaml/0.9/serde_yaml/with/singleton_map/index.html on whichever field you want the opposite representation for.

**Comment by Eric-Arellano** [comment]
Thanks for the workaround, @dtolnay! That worked for me :)

**Comment by willfindlay** [comment]
@dtolnay you're a legend, thank you. Is this documented anywhere and I just missed it? I would suggest making it more obvious unless I'm just blind.

**Comment by willfindlay** [comment]
@dtolnay Follow-up question: What would be the easiest/best way to make this work with a `Vec<T>` for example where `T` is an untagged enum?

Also, wouldn't it have been better to make this new behaviour opt-in? To me, it feels like wanting to use `!Tag` would be the exception rather than the rule considering so many use cases require compatibility with JSON.

**Comment by kaikalii** [comment]
> @dtolnay Follow-up question: What would be the easiest/best way to make this work with a `Vec<T>` for example where `T` is an untagged enum?
> 
> Also, wouldn't it have been better to make this new behaviour opt-in? To me, it feels like wanting to use `!Tag` would be the exception rather than the rule considering so many use cases require compatibility with JSON.

I second this. Could this be safely reverted?

Perhaps their should be some kind of settings struct that lets you choose how you want it to work?

**Comment by willfindlay** [comment]
> Perhaps there should be some kind of settings struct that lets you choose how you want it to work?

Or maybe even just a feature flag 👀

**Comment by divergentdave** [comment]
A feature flag would be unwise. In a diamond dependency scenario, crates would change each other's  format by setting that feature flag. serde_yaml would get compiled once, with the flag set, to be used by both its dependents.

**Comment by willfindlay** [comment]
Ah good point @divergentdave, makes sense.

**Comment by LGUG2Z** [comment]
> You would need to use https://docs.rs/serde_yaml/0.9/serde_yaml/with/singleton_map/index.html on whichever field you want the opposite representation for.

This doesn't appear to work if you apply it to a struct with an `Option<Enum>`; the `with` attribute somehow wants to treat this as a non-optional and fails the parsing when the `Option<Enum>` value is not included in the `yaml` file. Perhaps I'm missing something?

**Comment by sj4nes** [comment]
This is stumbling me today, I want to generate some YAML for Github Actions "Workflow", the `on:` field. Example:
```yaml
on:
  label:
    types: [opened]
```
with ```rust
#[derive(Serialize, Deserialize)]
pub enum LabelTypes {
    #[serde(rename = "opened")]
    Opened,
    #[serde(rename = "labeled")]
    Labeled,
}

#[derive(Serialize, Deserialize)]
#[allow(non_camel_case_types)]
pub enum GHAEvent {
    #[serde(rename = "push")]
    Push,
    #[serde(with = "serde_yaml::with::singleton_map")]
    label { types: Vec<LabelTypes> },
}
```
With some extra weirdness to get the label not be uppercased "Label" but I'm still getting ```yaml
on:
- !label
  - opened
```
adding `rename = "label"` to the serde annotation and leaving out the `#[allow(non_camel_case_types)]` still leaves me with a tagged `!label` list. I'm ignoring the inline vs. lined rendering since they mean the same thing.

Round tripping from YAML->Serde would be nice, but I'm not expecting it, I'm more focused on typed checked data structures doing the work and serde_yaml writing the YAML for me.

**Comment by dimonomid** [comment]
I know this issue is closed already, but for anyone coming across and looking for the answer on this:

> What would be the easiest/best way to make this work with a `Vec<T>` for example where `T` is an untagged enum?

Apparently the solution is to use `#[serde(with = "serde_yaml::with::singleton_map_recursive")]` (note the `_recursive` part). With this, all the nested types will use the singleton map approach.

And re: this:

> Also, wouldn't it have been better to make this new behaviour opt-in? To me, it feels like wanting to use `!Tag` would be the exception rather than the rule considering so many use cases require compatibility with JSON.

Definitely agree with that, because:

- The old tried-and-true singleton map approach wasn't broken. Don't fix what's not broken!
- With 0.9.25 (the latest version as of writing this), new tags approach is actually kinda broken! (nested enums aren't even supported)
- Changing the default is a breaking change, and breaking changes in general are painful, so there should be some really strong justification for it. So e.g. if (1) the whole industry was using yaml tags heavily, and (2) the implementation of it was opt-in for some time and has proven to be working well in all scenarios that the old implementation covers... _Then_ it could have been justified. But clearly it's not the case. Yaml tags are _not_ heavily used; in fact it remains pretty much an obscure feature of yaml which few people are even aware about. And as pointed out above, as of the current latest version, the implementation is subpar, since not all cases are supported.

So bottom line, we replace a good working solution with a half-broken one, and make it a default, without even clear docs on how to get the old behavior (Imo the release notes should clarify it in detail, with all those attributes). Overall, this seems like a bad call to me. Alas.
